exchanges/binance/binance_websocket.go

```diff
@@ -296,158 +299,127 @@ func (b *Binance) wsHandleData(respRaw []byte) error {
+				if !saveTradeData &&
+					!b.IsTradeFeedEnabled() {
+					return nil
 				}
 
-				pairs, err := b.GetEnabledPairs(asset.Spot)
+				var t TradeStream
+				err = json.Unmarshal(jsonData, &t)
 				if err != nil {
-					return err
+					return fmt.Errorf("%v - Could not unmarshal trade data: %s",
+						b.Name,
+						err)
 				}
-
-				format, err := b.GetPairFormat(asset.Spot, true)
+				return b.Websocket.Trade.Update(saveTradeData,
+					trade.Data{
+						CurrencyPair: pair,
+						Timestamp:    t.TimeStamp,
+						Price:        t.Price.Float64(),
+						Amount:       t.Quantity.Float64(),
+						Exchange:     b.Name,
+						AssetType:    asset.Spot,
+						TID:          strconv.FormatInt(t.TradeID, 10),
+					})
+			case "ticker":
+				var t TickerStream
+				err = json.Unmarshal(jsonData, &t)
 				if err != nil {
-					return err
+					return fmt.Errorf("%v - Could not convert to a TickerStream structure %s",
+						b.Name,
+						err.Error())
 				}
-
-				switch streamType[1] {
-				case "trade":
-					saveTradeData := b.IsSaveTradeDataEnabled()
-
-					if !saveTradeData &&
-						!b.IsTradeFeedEnabled() {
+				b.Websocket.DataHandler <- &ticker.Price{
+					ExchangeName: b.Name,
+					Open:         t.OpenPrice.Float64(),
+					Close:        t.ClosePrice.Float64(),
+					Volume:       t.TotalTradedVolume.Float64(),
+					QuoteVolume:  t.TotalTradedQuoteVolume.Float64(),
+					High:         t.HighPrice.Float64(),
+					Low:          t.LowPrice.Float64(),
+					Bid:          t.BestBidPrice.Float64(),
+					Ask:          t.BestAskPrice.Float64(),
+					Last:         t.LastPrice.Float64(),
+					LastUpdated:  t.EventTime,
+					AssetType:    asset.Spot,
+					Pair:         pair,
+				}
+				return nil
+			case "kline_1m", "kline_3m", "kline_5m", "kline_15m", "kline_30m", "kline_1h", "kline_2h", "kline_4h",
+				"kline_6h", "kline_8h", "kline_12h", "kline_1d", "kline_3d", "kline_1w", "kline_1M":
+				var kline KlineStream
+				err = json.Unmarshal(jsonData, &kline)
+				if err != nil {
+					return fmt.Errorf("%v - Could not convert to a KlineStream structure %s",
+						b.Name,
+						err)
+				}
+				b.Websocket.DataHandler <- stream.KlineData{
+					Timestamp:  kline.EventTime,
+					Pair:       pair,
+					AssetType:  asset.Spot,
+					Exchange:   b.Name,
+					StartTime:  kline.Kline.StartTime,
+					CloseTime:  kline.Kline.CloseTime,
+					Interval:   kline.Kline.Interval,
+					OpenPrice:  kline.Kline.OpenPrice.Float64(),
+					ClosePrice: kline.Kline.ClosePrice.Float64(),
+					HighPrice:  kline.Kline.HighPrice.Float64(),
+					LowPrice:   kline.Kline.LowPrice.Float64(),
+					Volume:     kline.Kline.Volume.Float64(),
+				}
+				return nil
+			case "depth":
+				var depth WebsocketDepthStream
+				err = json.Unmarshal(jsonData, &depth)
+				if err != nil {
+					return fmt.Errorf("%v - Could not convert to depthStream structure %s",
+						b.Name,
+						err)
+				}
+				var init bool
+				init, err = b.UpdateLocalBuffer(&depth)
```

Comment:
OOS suggestion: 

```diff
diff --git a/exchanges/binance/binance_websocket.go b/exchanges/binance/binance_websocket.go
index bf3c1aa40..697484fdb 100644
--- a/exchanges/binance/binance_websocket.go
+++ b/exchanges/binance/binance_websocket.go
@@ -485,35 +485,23 @@ func (b *Binance) SeedLocalCacheWithBook(p currency.Pair, orderbookNew *OrderBoo

 // UpdateLocalBuffer updates and returns the most recent iteration of the orderbook
 func (b *Binance) UpdateLocalBuffer(wsdp *WebsocketDepthStream) (bool, error) {
-       enabledPairs, err := b.GetEnabledPairs(asset.Spot)
+       pair, err := b.MatchSymbolWithAvailablePairs(wsdp.Pair, asset.Spot, false)
        if err != nil {
                return false, err
        }

-       format, err := b.GetPairFormat(asset.Spot, true)
+       err = b.obm.stageWsUpdate(wsdp, pair, asset.Spot)
        if err != nil {
-               return false, err
-       }
-
-       currencyPair, err := currency.NewPairFromFormattedPairs(wsdp.Pair,
-               enabledPairs,
-               format)
-       if err != nil {
-               return false, err
-       }
-
-       err = b.obm.stageWsUpdate(wsdp, currencyPair, asset.Spot)
-       if err != nil {
-               init, err2 := b.obm.checkIsInitialSync(currencyPair)
+               init, err2 := b.obm.checkIsInitialSync(pair)
                if err2 != nil {
                        return false, err2
                }
                return init, err
        }

-       err = b.applyBufferUpdate(currencyPair)
+       err = b.applyBufferUpdate(pair)
        if err != nil {
-               b.flushAndCleanup(currencyPair)
+               b.flushAndCleanup(pair)
        }

        return false, err
```

Reply:
I've got WS orderbook disabled, so that's probably why it didn't show up as a thing to look at. I'll have a look and comment later on. _So_ out of scope 🌞

Reply:
👍

Reply:
gr8 upd8 m8, 8/8